File: Projeto/app/src/main/java/com/example/fabio/bankapp/classes/CustomAdapter.java
package com.example.fabio.bankapp.classes;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.fabio.bankapp.R;

import java.text.DateFormat;
import java.text.NumberFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class CustomAdapter extends RecyclerView.Adapter<CustomAdapter.ViewHolder> {

    private Context context;
    private List<DataRecentes> recentesList;
    private Locale ptBr = new Locale("pt", "BR");

    public CustomAdapter(Context context, List<DataRecentes> recentesList) {
        this.context = context;
        this.recentesList = recentesList;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.recyclerview_recentes, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        String datastr, valorstr;
        Double valor;
        holder.txttitle.setText(recentesList.get(position).getTitle());
        holder.txtdesc.setText(recentesList.get(position).getDesc());

        datastr = recentesList.get(position).getDate();
        DateFormat formatUS = new SimpleDateFormat("yyyy-mm-dd");
        Date date = null;
        try {
            date = formatUS.parse(datastr);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        DateFormat formatBR = new SimpleDateFormat("dd/mm/yyyy");
        String dateFormated = formatBR.format(date);
        holder.txtdate.setText(dateFormated);

        valor = recentesList.get(position).getValue();
        valorstr = NumberFormat.getCurrencyInstance(ptBr).format(valor);
        holder.txtvalue.setText(valorstr);
    }

    @Override
    public int getItemCount() {
        return recentesList.size();
    }

    public  class ViewHolder extends  RecyclerView.ViewHolder{

        public TextView txttitle, txtdesc, txtdate, txtvalue;

        public ViewHolder(View itemView) {
            super(itemView);
            txttitle = itemView.findViewById(R.id.txtTitle);
            txtdesc = itemView.findViewById(R.id.txtDesc);
            txtdate = itemView.findViewById(R.id.txtDate);
            txtvalue = itemView.findViewById(R.id.txtValue);
        }
    }
}

File: app/src/main/java/com/example/ibm_test/component/CustomAdapter.kt
package com.example.ibm_test.component

import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

class CustomAdapter<T>(
    private var items: MutableList<T>,
    private val recyclerViewItem: AdapterItem<T>? = null,
    private val itemsViewHolder: (inflater: LayoutInflater, parent: ViewGroup, viewType: Int) -> RecyclerView.ViewHolder,
    private val viewType: (position: Int) -> Int? = { null }
) : RecyclerView.Adapter<CustomViewHolder<T>>() {

    @Suppress("UNCHECKED_CAST")
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CustomViewHolder<T> {
        val inflater: LayoutInflater = LayoutInflater.from(parent.context)
        return itemsViewHolder(inflater, parent, viewType) as CustomViewHolder<T>
    }

    override fun onBindViewHolder(holder: CustomViewHolder<T>, position: Int) {
        val item = items[position]

        holder.setup(item, recyclerViewItem)
        recyclerViewItem?.position(position)
    }

    override fun getItemViewType(position: Int): Int =
        if (viewType(position) == null) {
            if (items[position] is ProgressViewHolder.LoadMore)
                1
            else
                0
        } else
            viewType(position) ?: -1


    override fun getItemCount(): Int = items.size


    fun loadMoreData(list: MutableList<T>) {
        items.addAll(list)
        notifyDataSetChanged()
    }
}

interface AdapterItem<T> {
    fun onClickItem(it: T?)
    fun position(position: Int)
}


class ProgressViewHolder(itemView: View) : CustomViewHolder<ProgressViewHolder.LoadMore>(itemView) {
    override fun buildView(recyclerViewItem: AdapterItem<LoadMore>?) {}
    data class LoadMore(val isLoadMore: Boolean = true)
}

abstract class CustomViewHolder<T>(itemView: View) : RecyclerView.ViewHolder(itemView) {
    private var item: T? = null

    fun setup(item: T, recyclerViewItem: AdapterItem<T>?) {
        this.item = item
        buildView(recyclerViewItem)
    }


    fun get(): T? = item


    abstract fun buildView(recyclerViewItem: AdapterItem<T>?)
}